docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,247 @@
+- a checkpoint when installing CAPZ in the framework that ASO is also installed and healthy
+- upgrade tests from a version of CAPZ that does not use ASO to one that does,
+  verifying that CAPZ-managed and unmanaged Azure resources become correctly
+  reflected in ASO after the upgrade
+- a test to verify when a CAPZ upgrade includes an upgrade of the ASO control plane
+- a test to verify when a CAPZ upgrade includes a newer API version of an ASO resource
+
+To catch performance and scalability regressions, one-off tests against CAPZ instances managing many and/or large workload clusters will also be performed. Telemetry will be inspected to ensure the time it takes to reconcile a high volume of clusters and the overhead required in the management cluster to run ASO are still acceptable.
+
+Finally, unit tests will be added to cover tightly-scoped functional changes related to ASO adoption, like error handling.
+
+### Graduation Criteria
+
+ASO integration will not be kept behind a feature flag or matriculate through the usual alpha, beta, and stable phases. Instead, the transition will be made gradually so as to distribute potential impact over time.
+
+Each Azure service package, such as `virtualmachines` or `loadbalancers` can be transitioned separately from using the SDK to ASO. The transition for a package will be made in one atomic change, so the SDK- and ASO-based implementations will never exist side-by-side for a given service. However, there will be a period of time where the SDK and ASO are driving mutually exclusive sets of services at the same time, and that time span may include several releases of CAPZ. If significant regressions are found in a particular service implementation, that service (and others that rely on it) can be rolled back in isolation to the SDK-based implementation independently of unaffected services.
+
+Since the transition for each service is expected to take some time, it's worth considering which services to transition first. ASO's resources include some references to Azure resources as references to ASO resources which will impose some necessary ordering. For example, a VirtualNetworksSubnet's `spec.owner` refers to the Kubernetes name of the owning ASO VirtualNetwork, so the `virtualnetworks` service would be implemented to use ASO before the same is done for `subnets`. When no ordering is otherwise necessary, the priority to implement each service will be based on new features which are blocked on newer Azure API versions available in ASO but not the SDK and also community involvement.
```

Comment:
> ASO's resources include some references to Azure resources as references to ASO resources which will impose some necessary ordering.

I had to read this sentence several times but I still can't parse it. I understand the gist from the subsequent sentences: the dependencies between resources require us to implement some ASO resources before others.

Reply:
wow yes I turned that into quite the mess. I'll try again to make that comprehensible.